src/merkle_tree/mod.rs

```diff
@@ -1,253 +1,399 @@
+/// solve the problem.
 pub struct MerkleTree<P: Config> {
-    tree: Vec<Digest<P>>,
-    padding_tree: Vec<(Digest<P>, Digest<P>)>,
-    parameters: <P::H as FixedLengthCRH>::Parameters,
-    root: Option<Digest<P>>,
+    /// stores the non-leaf nodes in level order. The first element is the root node.
+    non_leaf_nodes: Vec<TwoToOneDigest<P>>,
+    /// store the hash of leaf nodes from left to right
+    leaf_nodes: Vec<LeafDigest<P>>,
+    /// Store the two-to-one hash parameters
+    two_to_one_hash_param: TwoToOneParam<P>,
+    /// Store the leaf hash parameters
+    leaf_hash_param: LeafParam<P>,
+    /// Stores the height of the MerkleTree
+    height: usize,
 }
 
 impl<P: Config> MerkleTree<P> {
-    pub const HEIGHT: u8 = P::HEIGHT as u8;
-
-    pub fn blank(parameters: Parameters<P>) -> Self {
-        MerkleTree {
-            tree: Vec::new(),
-            padding_tree: Vec::new(),
-            root: None,
-            parameters,
-        }
+    /// Create an empty merkle tree such that all leaves are zero-filled.
+    pub fn blank(
+        leaf_hash_param: &LeafParam<P>,
+        two_to_one_hash_param: &TwoToOneParam<P>,
+        height: usize,
+    ) -> Result<Self, crate::Error> {
+        let leaf = vec![0u8; P::LeafHash::INPUT_SIZE_BITS / 8];
+        let leaves = vec![leaf; 1 << (height - 1)];
+        Self::new(leaf_hash_param, two_to_one_hash_param, &leaves)
     }
 
-    pub fn new<L: ToBytes>(parameters: Parameters<P>, leaves: &[L]) -> Result<Self, crate::Error> {
-        let new_time = start_timer!(|| "MerkleTree::New");
-
-        let last_level_size = leaves.len().next_power_of_two();
-        let tree_size = 2 * last_level_size - 1;
-        let tree_height = tree_height(tree_size);
-        assert!(tree_height as u8 <= Self::HEIGHT);
-
-        // Initialize the merkle tree.
-        let mut tree = Vec::with_capacity(tree_size);
-        let empty_hash = hash_empty::<P::H>(&parameters)?;
-        for _ in 0..tree_size {
-            tree.push(empty_hash.clone());
-        }
-
-        // Compute the starting indices for each level of the tree.
+    /// Returns a new merkle tree. `leaves.len()` should be power of two.
+    pub fn new<L: ToBytes>(
+        leaf_hash_param: &LeafParam<P>,
+        two_to_one_hash_param: &TwoToOneParam<P>,
+        leaves: &[L],
+    ) -> Result<Self, crate::Error> {
+        let leaf_nodes_size = leaves.len(); // size of the leaf layer
+        assert!(
+            leaf_nodes_size.is_power_of_two(),
+            "`leaves.len() should be power of two"
+        );
+        let non_leaf_nodes_size = leaf_nodes_size - 1;
+
+        let tree_height = tree_height(non_leaf_nodes_size + leaf_nodes_size);
+
+        let hash_of_empty: TwoToOneDigest<P> = P::TwoToOneHash::evaluate_both(
+            two_to_one_hash_param,
+            &vec![0u8; P::TwoToOneHash::LEFT_INPUT_SIZE_BITS / 8],
```

Comment:
Shouldn't this be `(P::TwoToOneHash::LEFT_INPUT_SIZE_BITS + 7) / 8` unless the number of bits is guaranteed to be a multiple of 8?

Reply:
Wouldn't it overflow? For example, if LEFT_INPUT_SIZE_BITS is 15 bits, then `(P::TwoToOneHash::LEFT_INPUT_SIZE_BITS + 7) / 8`  is 2 bytes which is 16bits

Reply:
If it wanted 15 bits of input, wouldn't that have to be provided as 2 bytes, and it just ignores 1 bit?

Reply:
In existing implementations, there is an assert that enforces number of bytes * 8 < INPUT_SIZE_BITS, so I'm not sure if we can release the constraints:
https://github.com/arkworks-rs/crypto-primitives/blob/1a71386a10ef627834e4c3b14ca982811f23192d/src/crh/pedersen/mod.rs#L70-L80

Anyway, `hash_of_empty` is just a dummy placeholder which will not be stored in actual merkle tree. We might need to modify the existing CRH implementations release that constraints...

Reply:
oh interesting, I had no idea. If that assert is there, I don't understand how it works with non-multiple of 8 window sizes. Sounds good for us to file this as an issue to revisit once the MT is merged.

Reply:
Bytes are always 8-bit sized, so I don't see how to avoid this? That is, there will always be some wasted space when using bytes as the input.